src/tools/kata-ctl/src/ops/volume_ops.rs

```diff
@@ -0,0 +1,294 @@
+        // turn file_name into String and return it
+        let file_name = file_name.to_str().ok_or_else(|| {
+            anyhow!(
+                "failed to convert file_name {:?} to string",
+                file_name.to_string_lossy()
+            )
+        })?;
+
+        return Ok(String::from(file_name));
+    }
+
+    return Err(anyhow!("no sandbox found for {}", volume_path));
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use kata_types::mount::DirectVolumeMountInfo;
+    use std::{collections::HashMap, fs};
+    use tempfile::tempdir;
+    use test_utils::skip_if_not_root;
+
+    #[test]
+    fn test_get_sandbox_id_for_volume() {
+        // this test has to run as root, so has to manually cleanup afterwards
+        skip_if_not_root!();
+
+        // create KATA_DIRECT_VOLUME_ROOT_PATH first as safe_path::scoped_join
+        // requires prefix dir to exist
+        fs::create_dir_all(KATA_DIRECT_VOLUME_ROOT_PATH)
+            .expect("create kata direct volume root path failed");
+
+        let test_sandbox_id = "sandboxid_test_file";
+        let test_volume_path = String::from("a/b/c");
+        let joined_volume_path =
+            join_path(KATA_DIRECT_VOLUME_ROOT_PATH, &test_volume_path).unwrap();
+
+        let test_file_dir = joined_volume_path.join(test_sandbox_id);
+        fs::create_dir_all(&joined_volume_path).expect("failed to mkdir -p");
+        fs::write(&test_file_dir, "teststring").expect("failed to write");
+
+        // test that get_sandbox_id gets the correct sandboxid it sees
+        let got = get_sandbox_id_for_volume(&test_volume_path).unwrap();
+        assert!(got.eq(test_sandbox_id));
+
+        // test that get_sandbox_id returns error if no sandboxid found
+        fs::remove_file(&test_file_dir).expect("failed to remove");
+        get_sandbox_id_for_volume(&test_volume_path).expect_err("error expected");
+
+        // cleanup test directory
+        fs::remove_dir_all(&joined_volume_path).expect("failed to cleanup test")
+    }
+
+    #[test]
+    fn test_path_join() {
+        #[derive(Debug)]
+        struct TestData<'a> {
+            rootfs: &'a str,
+            volume_path: &'a str,
+            result: Result<PathBuf>,
+        }
+        // the safe_path::scoped_join requires the prefix path to exist on testing machine
+        let root_fs = tempdir().expect("failed to create tmpdir").into_path();
+        let root_fs_str = root_fs.to_str().unwrap();
+
+        // this b64encodes to "/abcdddd"
+        let encoded_b64_abc = vec![253, 166, 220, 117, 215, 93];
+
+        let tests = &[
+            TestData {
+                rootfs: root_fs_str,
+                volume_path: "",
+                result: Err(anyhow!("volume path must not be empty")),
+            },
+            TestData {
+                rootfs: root_fs_str,
+                volume_path: "../../etc/passwd",
+                result: Ok(root_fs.join("Li4vLi4vZXRjL3Bhc3N3ZA==")),
```

Comment:
This is fine, but for clarity I'd consider creating a few variables to hold the original string version of `volume_path`, then calling `base64::encode()` so that the only visible strings are the human-readable ones. Something like:

```rust
let relative_passwd_path = "../../etc/passwd";
let b64_relative_passwd_path = base64::encode(relative_passwd_path.clone());

// ...

let tests = &[
    TestData {
        rootfs: root_fs_str,
        volume_path: relative_passwd_path,
        result: Ok(root_fs.join(b64_relative_passwd_path),
    }

    // ...
```